controller/kubernetes_pod_controller.go

```diff
@@ -0,0 +1,271 @@
+	kc.queue.Forget(key)
+}
+
+func (kc *KubernetesPodController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, controllerAgentName+": fail to sync %v", key)
+	}()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	pod, err := kc.pLister.Pods(namespace).Get(name)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "Error getting Pod: %s", name)
+	}
+	nodeID := pod.Spec.NodeName
+	kc.forceDeletePod(pod, nodeID, namespace)
+
+	return nil
+}
+
+// Force delete a pod when all of the below conditions are meet:
+// 1. StatefulSetPodDeletePolicy is StatefulSetPodDeletePolicyForceDelete
+// 2. pod is belonging to a StatefulSet
+// 3. node containing the pod is downed
+// 4. the pod is terminating
+// 5. pod has a PV with provisioner driver.longhorn.io
+func (kc *KubernetesPodController) forceDeletePod(pod *v1.Pod, nodeID string, namespace string) {
+	// by default we don't force delete pod of StatefulSet
+	var deletionPolicy types.StatefulSetPodDeletePolicy
+	if deletionSetting, err := kc.ds.GetSettingValueExisted(types.SettingNameStatefulSetPodDeletePolicy); err != nil {
+		deletionPolicy = types.StatefulSetPodDeletePolicyNever
+	} else {
+		deletionPolicy = types.StatefulSetPodDeletePolicy(deletionSetting)
+	}
+
+	if deletionPolicy == types.StatefulSetPodDeletePolicyNever {
+		return
+	}
+
+	if !isOwnedByStatefulSet(pod) {
+		return
+	}
+
+	// only delete StatefulSet pods on the node which is declared `NotReady` or doesn't exist.
+	isNodeDowned, err := kc.ds.IsNodeDownOrDeleted(nodeID)
+	if err != nil {
+		logrus.Errorf("failed to evaluate Node %v for pod %v in forceDeletePod: %v", nodeID, pod.Name, err)
+		return
+	}
+	if !isNodeDowned {
+		return
+	}
+
+	if pod.DeletionTimestamp != nil && pod.DeletionTimestamp.After(time.Now()) {
+		gracePeriod := int64(0)
+		err = kc.kubeClient.CoreV1().Pods(namespace).Delete(pod.Name, &metav1.DeleteOptions{
+			GracePeriodSeconds: &gracePeriod,
+		})
+		if err != nil {
+			logrus.Errorf("failed to forcefully delete StatefulSet Pod %v on the downed Node %v in forceDeletePod: %v", pod.Name, nodeID, err)
+			return
+		}
+		kc.eventRecorder.Eventf(pod, v1.EventTypeNormal, EventReasonDelete, "Forcefully deleted StatefulSet Pod %v on 'NotReady' Node %v", pod.Name, nodeID)
+	}
+
+	return
+}
+
+func isOwnedByStatefulSet(pod *v1.Pod) bool {
+	if ownerRef := metav1.GetControllerOf(pod); ownerRef != nil {
+		return ownerRef.Kind == types.KubernetesStatefulSet
+	}
+	return false
+}
+
+// We only handle pods which have PV with storage provisioner is driver.longhorn.io
+func (kc *KubernetesPodController) enqueuePodChange(pod *v1.Pod) {
+	key, err := controller.KeyFunc(pod)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", pod, err))
+		return
+	}
+
+	for _, v := range pod.Spec.Volumes {
+		if v.VolumeSource.PersistentVolumeClaim != nil {
```

Comment:
consider inverting & continue
that way you have all your preconditions checked before and in the same way.
this also reduces the nesting by 1 level

Reply:
Thank you. Fixed